diff --git a/platform/javascript/detect.py b/platform/javascript/detect.py
old mode 100644
new mode 100644
index b527336f83..60af313de0
--- a/platform/javascript/detect.py
+++ b/platform/javascript/detect.py
@@ -98,7 +98,7 @@ def configure(env):
     else:
         # Disable exceptions and rtti on non-tools (template) builds
         # These flags help keep the file size down.
-        env.Append(CCFLAGS=["-fno-exceptions", "-fno-rtti"])
+        env.Append(CCFLAGS=["-fno-exceptions"])
         # Don't use dynamic_cast, necessary with no-rtti.
         env.Append(CPPDEFINES=["NO_SAFE_CAST"])
 
