diff --git a/editor/code_editor.cpp b/editor/code_editor.cpp
index 95c7cc0bf1..5a1f8d7b2d 100644
--- a/editor/code_editor.cpp
+++ b/editor/code_editor.cpp
@@ -1734,7 +1734,7 @@ CodeTextEditor::CodeTextEditor() {
 	error->set_v_size_flags(SIZE_EXPAND | SIZE_SHRINK_CENTER);
 	error->set_mouse_filter(MOUSE_FILTER_STOP);
 	error->connect("gui_input", callable_mp(this, &CodeTextEditor::_error_pressed));
-	find_replace_bar->connect("error", callable_mp(error, &Label::set_text));
+	find_replace_bar->connect("error", callable_mp((TLLabel *)error, &TLLabel::set_text));
 
 	// Warnings
 	warning_button = memnew(Button);
