diff --git a/SConstruct b/SConstruct
index 1f06a0a..ba80f6a 100644
--- a/SConstruct
+++ b/SConstruct
@@ -386,7 +386,8 @@ elif env['platform'] == 'android':
     # Setup tools
     env['CC'] = toolchain + "/bin/clang"
     env['CXX'] = toolchain + "/bin/clang++"
-    env['AR'] = toolchain + "/bin/" + arch_info['tool_path'] + "-ar"
+    env['AR'] = toolchain + "/bin/llvm-ar"
+    env['RANLIB'] = toolchain + "/bin/llvm-ranlib"
 
     env.Append(CCFLAGS=['--target=' + arch_info['target'] + env['android_api_level'], '-march=' + arch_info['march'], '-fPIC'])#, '-fPIE', '-fno-addrsig', '-Oz'])
     env.Append(CCFLAGS=arch_info['ccflags'])
